cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1677,6 +1703,177 @@ struct GroupedMinMaxImpl final : public GroupedAggregator {
+               const ArrayData& group_id_mapping) override {
+    auto other = checked_cast<GroupedMinMaxImpl*>(&raw_other);
+    auto g = group_id_mapping.GetValues<uint32_t>(1);
+    for (uint32_t other_g = 0; static_cast<int64_t>(other_g) < group_id_mapping.length;
+         ++other_g, ++g) {
+      if (!mins_[*g] ||
+          (mins_[*g] && other->mins_[other_g] &&
+           util::string_view(*mins_[*g]) > util::string_view(*other->mins_[other_g]))) {
+        mins_[*g] = std::move(other->mins_[other_g]);
+      }
+      if (!maxes_[*g] ||
+          (maxes_[*g] && other->maxes_[other_g] &&
+           util::string_view(*maxes_[*g]) < util::string_view(*other->maxes_[other_g]))) {
+        maxes_[*g] = std::move(other->maxes_[other_g]);
+      }
+
+      if (BitUtil::GetBit(other->has_values_.data(), other_g)) {
+        BitUtil::SetBit(has_values_.mutable_data(), *g);
+      }
+      if (BitUtil::GetBit(other->has_nulls_.data(), other_g)) {
+        BitUtil::SetBit(has_nulls_.mutable_data(), *g);
+      }
+    }
+    return Status::OK();
+  }
+
+  Result<Datum> Finalize() override {
+    // aggregation for group is valid if there was at least one value in that group
+    ARROW_ASSIGN_OR_RAISE(auto null_bitmap, has_values_.Finish());
+
+    if (!options_.skip_nulls) {
+      // ... and there were no nulls in that group
+      ARROW_ASSIGN_OR_RAISE(auto has_nulls, has_nulls_.Finish());
+      arrow::internal::BitmapAndNot(null_bitmap->data(), 0, has_nulls->data(), 0,
+                                    num_groups_, 0, null_bitmap->mutable_data());
+    }
+
+    auto mins = ArrayData::Make(type_, num_groups_, {null_bitmap, nullptr});
+    auto maxes = ArrayData::Make(type_, num_groups_, {std::move(null_bitmap), nullptr});
+    RETURN_NOT_OK(MakeOffsetsValues(mins.get(), mins_));
+    RETURN_NOT_OK(MakeOffsetsValues(maxes.get(), maxes_));
+    return ArrayData::Make(out_type(), num_groups_, {nullptr},
+                           {std::move(mins), std::move(maxes)});
+  }
+
+  template <typename T = Type>
+  enable_if_base_binary<T, Status> MakeOffsetsValues(
+      ArrayData* array, const std::vector<std::shared_ptr<ResizableBuffer>>& values) {
+    using offset_type = typename T::offset_type;
+    ARROW_ASSIGN_OR_RAISE(
+        auto raw_offsets,
+        AllocateBuffer((1 + values.size()) * sizeof(offset_type), ctx_->memory_pool()));
+    offset_type* offsets = reinterpret_cast<offset_type*>(raw_offsets->mutable_data());
+    offsets[0] = 0;
+    offsets++;
+    const uint8_t* null_bitmap = array->buffers[0]->data();
+    int64_t total_length = 0;
+    for (size_t i = 0; i < values.size(); i++) {
+      if (BitUtil::GetBit(null_bitmap, i)) {
+        total_length += values[i]->size();
+      }
+      if (total_length > std::numeric_limits<offset_type>::max()) {
+        return Status::Invalid("Result is too large to fit in ", *array->type,
+                               " cast to large_ variant of type");
```

Comment:
If `offset_type` is of type `int64_t` then this limit check will never trigger and will allow overflow:
```c++
if (BitUtil::GetBit(null_bitmap, i)) {
    auto value_size = values[i]->size();
    if ((total_length > 0) && (value_size > (std::numeric_limits<offset_type>::max() - total_length)) {
        return Status::Invalid(...);
    }
    total_length += value_size;
}
```

Or use [safe-math.h utils](https://github.com/apache/arrow/blob/master/cpp/src/arrow/vendored/portable-snippets/safe-math.h).

Reply:
Frankly we're awful about checking for overflow in general; most kernel code is like this and only checks for the case of overflowing int32_t. That said I'll add the check here.

Reply:
Added an overflow check.